neuro_evolution_ctrnn/brains/i_brain.py

```diff
@@ -66,69 +76,111 @@ def get_class_state(cls):
+                                 relevant_space: spaces.Space = None) -> Union[np.ndarray, int, Tuple, list]:
+        """
+        Parses 'brain_input_output' to a new 'format'. If is_brain_input is True this is a np.ndarray. If is_brain_input
+        is False the format corresponds to the output Space, i.e. the action space of the environment.
+
+        An input for the brain equals the output of an environment and vice versa.
+
+        Discrete Space as input space will be encoded as a one hot vector, Discrete Space as output takes the argmax.
+        Box Space as input will not be transformed, Box Space as output will be reshaped to match the environment.
+        Tuple Space as input unpacks each subspace, transforms it with this function (recursive) and then flattens every
+        output of the subspaces. For Tuple Space as output a tuple will be created from the brain output where the
+        entries are of the correct shape for the environment.
+
+        Note that Tuple Spaces inside Tuple Spaces are not supported.
+
+        :param brain_input_output: The to be parsed variable
+        :param is_brain_input: Should be True if brain_input_output is the input for a Brain, False if not
+        :param relevant_space: relevant_space is normally chosen depending on is_brain_input but it can be provided
+        directly with this parameter. This is used for Tuple spaces which are nested
+        :return: The parsed variable
+        """
+        # This is the Space we need to calculate the correct output form
+        if relevant_space is None:
+            relevant_space = self.input_space if is_brain_input else self.output_space
+
+        if isinstance(relevant_space, spaces.Box):
+            if is_brain_input:
+                return brain_input_output
+            else:
+                return brain_input_output.reshape(relevant_space.shape)
+        elif isinstance(relevant_space, spaces.Discrete):
+            # We encode Discrete Spaces as one-hot vectors
+            if is_brain_input:
+                one_hot_vector = np.zeros(relevant_space.n)
+                one_hot_vector[brain_input_output] = 1
+                return one_hot_vector
+            else:
+                return np.argmax(brain_input_output)
+        elif isinstance(relevant_space, spaces.Tuple):
+            # Tuple Spaces have a tuple of "nested" Spaces. Environment observations (inputs for the brain) is therefore
+            # a tuple where each entry in the tuple is linked to the corresponding Space.
+            if is_brain_input:
+                # Transform the brain input (a tuple) into a list which is structured to resemble the "nested" Spaces.
+                brain_input = []
+                for i, sub_input in enumerate(brain_input_output):
+                    if isinstance(relevant_space[i], spaces.Tuple):
+                        raise RuntimeError("""Gym Tuple Spaces inside a gym Tuple space are not supported, maybe choose
```

Comment:
Yes but that's causing problems in some brains. Consider a space `Tuple(Tuple(Box, Box), Tuple(Discrete))`. Then the resulting input would be something like  `[[Box_parsed, Box_parsed], [Discrete_parsed]]` so a two-dimensional input for the brain. This could again be flattened but I don't know if this is really desired - some thinking has to go into if this makes actually sense from a standpoint where you think about the environment and the training. Besides I don't know if there are actually environments with nested Tuple Spaces and if there are then the error is thrown to force us into thinking about it.